src/client/Microsoft.Identity.Client/Internal/Requests/CredentialBasedMsiAuthRequest.cs

```diff
@@ -0,0 +1,235 @@
+                {
+                    _requestContext.Logger.Error("[Managed Identity] Response is either null or insufficient for authentication.");
+                    throw new MsalManagedIdentityException(
+                        MsalError.ManagedIdentityRequestFailed,
+                        MsalErrorMessage.ManagedIdentityInvalidResponse,
+                        _sourceType);
+                }
+
+                return credentialResponse;
+            }
+
+            private static string CreateCredentialPayload(X509Certificate2 x509Certificate2)
+            {
+                string certificateBase64 = Convert.ToBase64String(x509Certificate2.Export(X509ContentType.Cert));
+
+                return @"
+                    {
+                        ""cnf"": {
+                            ""jwk"": {
+                                ""kty"": ""RSA"", 
+                                ""use"": ""sig"",
+                                ""alg"": ""RS256"",
+                                ""kid"": """ + x509Certificate2.Thumbprint + @""",
+                                ""x5c"": [""" + certificateBase64 + @"""]
+                            }
+                        },
+                        ""latch_key"": false    
+                    }";
+            }
+        }
+
+        private readonly AcquireTokenForManagedIdentityParameters _managedIdentityParameters;
+        private CredentialSource _slc;
+        public static CredentialBasedMsiAuthRequest TryCreate(
+            IServiceBundle serviceBundle,
+            AuthenticationRequestParameters authenticationRequestParameters,
+            AcquireTokenForManagedIdentityParameters managedIdentityParameters)
+        {
+            // TODO: need static cahing on the state of SLC on the machine. Here or deeper down?
+            CredentialSource slc =
+                CredentialSource.TryCreate(authenticationRequestParameters.RequestContext);
+
+            if (slc != null)
+            {
+                return new CredentialBasedMsiAuthRequest(
+                    serviceBundle,
+                    authenticationRequestParameters,
+                    managedIdentityParameters);
+            }
+
+            return null;
+        }
+
+        private CredentialBasedMsiAuthRequest(
+            IServiceBundle serviceBundle,
+             AuthenticationRequestParameters authenticationRequestParameters,
+            AcquireTokenForManagedIdentityParameters managedIdentityParameters)
+            : base(serviceBundle, authenticationRequestParameters, managedIdentityParameters)
+        {
+            _managedIdentityParameters = managedIdentityParameters;
+        }
+
+        protected override Task<AuthenticationResult> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            // TODO: don't rely on AbstractManagedIdentity for SLC, we need to use ConfidentialClientApplication
```

Comment:
This is where all the SLC logic can be orchestrated. At this point `CredentialSource.TryCreate` has ran.

- extract authority, tenant etc. from SLC endpoint
- extract claims, client capabilities, force refresh etc., (all APIs) from the config.
- Create CCA and call `AcquireTokenForClient`